### PR TITLE
AI_CompareDamagingMoves changes to increase the score by 1 for best moves 

### PR DESCRIPTION
**Pull request by AlexOn1ine**
Adds temporary scores inside`AI_CompareDamagingMoves` have been added to that are decreased depending on the move. Once a good move has been found the score will be increased by 1.

The reason for these changes are status moves. They usually don't go as high and the AI might have started to ignore them even if they are beneficial.
  
Before:
image

After:
pokeemerald-2

Each good move gets a +1:
pokeemerald-3

### CODE REVIEW

File: src/battle_ai_main.c [hunk 1]
```diff
@@ -3169,8 +3169,9 @@ static u32 GetAIMostDamagingMoveId(u32 battlerAtk, u32 battlerDef)
 
 static s32 AI_CompareDamagingMoves(u32 battlerAtk, u32 battlerDef, u32 currId)
 {
-    u32 i;
+    u32 i, bestViableMoveSc0re;
```
[the remark this answers is not in the document]
Comment (DizzyEggg) [reply src/battle_ai_main.c:3172]: Also s32

File: src/battle_ai_main.c [hunk 2]
```diff
@@ -3187,11 +3188,13 @@ static s32 AI_CompareDamagingMoves(u32 battlerAtk, u32 battlerDef, u32 currId)
                 leastHits = noOfHits[i];
                 leastHitsId = i;
             }
+            viableMoveScores[i] = 100;
```
Comment (DizzyEggg) [inline src/battle_ai_main.c:3191]: viableMoveScores[i] = AI_SCORE_DEFAULT;

File: test/battle/ai.c [hunk 3]
```diff
@@ -105,6 +108,9 @@ AI_SINGLE_BATTLE_TEST("AI prefers moves with better accuracy, but only if they b
         ASSUME(gBattleMoves[MOVE_MEGA_KICK].accuracy < gBattleMoves[MOVE_STRENGTH].accuracy);
         ASSUME(gBattleMoves[MOVE_TACKLE].accuracy == 100);
         ASSUME(gBattleMoves[MOVE_GUST].accuracy == 100);
+        ASSUME(gBattleMoves[MOVE_SHOCK_WAVE].accuracy == 0);
+        ASSUME(gBattleMoves[MOVE_THUNDERBOLT].accuracy == 100);
+        ASSUME(gBattleMoves[MOVE_ICY_WIND].accuracy == 95);
```
Comment (DizzyEggg) [inline test/battle/ai.c:113]: Perhaps better to do ASSUME(gBattleMoves[MOVE_ICY_WIND].accuracy != 100);

File: test/battle/ai.c [hunk 4]
```diff
@@ -192,6 +198,65 @@ AI_SINGLE_BATTLE_TEST("AI prefers Earthquake over Drill Run if both require the
...
+    }
+}
+
+AI_SINGLE_BATTLE_TEST("AI can choose a stat boosting move randomly")
+{
+    GIVEN {
+        AI_FLAGS(AI_FLAG_CHECK_BAD_MOVE | AI_FLAG_CHECK_VIABILITY | AI_FLAG_TRY_TO_FAINT);
+        PLAYER(SPECIES_WOBBUFFET);
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_WOBBUFFET) { Moves(MOVE_THUNDERBOLT, MOVE_SLUDGE_BOMB, MOVE_SWORDS_DANCE); }
```
Comment (DizzyEggg) [inline test/battle/ai.c:254]: Actually, in this case it shouldn't ever choose Swords Dance, because both Tbold and Sludge Bomb are special moves.

**Review by DizzyEggg — CHANGES_REQUESTED** [review]

**Comment by DizzyEggg** [comment]
Nice to see you writing ai tests :D

File: test/battle/ai.c (hunk 4 above)
Comment (AlexOn1ine) [reply test/battle/ai.c:254]: True. Seems like a Sword Dance issue. I will add the check.

Comment (AlexOn1ine) [reply test/battle/ai.c]: I changed the test to highlight Sword Dance a little bit more.

File: test/battle/ai.c (hunk 3 above)
Comment (AlexOn1ine) [reply test/battle/ai.c]: fixed

File: src/battle_ai_main.c (hunk 2 above)
Comment (AlexOn1ine) [reply src/battle_ai_main.c]: fixed

File: src/battle_ai_main.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexOn1ine) [reply src/battle_ai_main.c]: applied

**Review by DizzyEggg — APPROVED** [review]

**Comment by AsparagusEduardo** [comment]
This PR doesn't compile in upcoming due to `move3` and `move4` not being used in `"AI prefers a weaker move over a one with a downside effect if both require the same number of hits to ko"`.

**Comment by DizzyEggg** [comment]
How did CI not catch that?

**Comment by AsparagusEduardo** [comment]
> How did CI not catch that?

I have no idea. I got the error trying to merge in upcoming to #3390 .

**Comment by AsparagusEduardo** [comment]
Ok, CI did catch it, but the CI that ran this PR was before we switched to modern I think
image

**Comment by DizzyEggg** [comment]
I see, well, should be an easy fix at least

**Comment by AsparagusEduardo** [comment]
Yeah. For now, let's rerun all PRs aimed at upcoming.

**Comment by AsparagusEduardo** [comment]
Wait no!
Let's rerun it once it's fixed.

**Comment by DizzyEggg** [comment]
Yeah, good idea.

**Pull request merged**